File: frontend/src/services/aiModel.ts
/**
 * AI 模型管理 API 服务
 */

import { api } from './api'

// 模型提供商枚举
export type ModelProvider = 
  | 'openai' 
  | 'openai_compatible' 
  | 'anthropic' 
  | 'stable_diffusion' 
  | 'comfyui' 
  | 'midjourney'

// 模型类型枚举
export type ModelType = 'text_generation' | 'image_generation' | 'image_analysis'

// 模型配置接口
export interface AIModel {
  id: string
  name: string
  provider: ModelProvider
  model_type: ModelType
  model_name: string
  
  // 连接配置
  base_url?: string
  api_version?: string
  organization_id?: string
  deployment_name?: string
  has_api_key: boolean
  
  // 文本生成配置
  default_max_tokens?: number
  default_temperature?: number
  
  // 图像生成配置
  default_width?: number
  default_height?: number
  default_steps?: number
  default_cfg_scale?: number
  default_sampler?: string
  checkpoint_name?: string
  workflow_template?: string
  
  // 通用配置
  timeout: number
  extra_headers?: Record<string, string>
  extra_config?: Record<string, any>
  
  is_enabled: boolean
  is_default: boolean
  created_at: string
  updated_at: string
}

// 创建模型配置请求
export interface AIModelCreateRequest {
  name: string
  provider: ModelProvider
  model_type: ModelType
  model_name: string
  
  api_key?: string
  base_url?: string
  api_version?: string
  organization_id?: string
  deployment_name?: string
  
  default_max_tokens?: number
  default_temperature?: number
  
  default_width?: number
  default_height?: number
  default_steps?: number
  default_cfg_scale?: number
  default_sampler?: string
  checkpoint_name?: string
  workflow_template?: string
  
  timeout?: number
  extra_headers?: Record<string, string>
  extra_config?: Record<string, any>
  
  is_enabled?: boolean
  is_default?: boolean
}

// 更新模型配置请求
export interface AIModelUpdateRequest {
  name?: string
  model_name?: string
  api_key?: string
  base_url?: string
  api_version?: string
  organization_id?: string
  deployment_name?: string
  
  default_max_tokens?: number
  default_temperature?: number
  
  default_width?: number
  default_height?: number
  default_steps?: number
  default_cfg_scale?: number
  default_sampler?: string
  checkpoint_name?: string
  workflow_template?: string
  
  timeout?: number
  extra_headers?: Record<string, string>
  extra_config?: Record<string, any>
  
  is_enabled?: boolean
  is_default?: boolean
}

// 模型列表响应
export interface AIModelListResponse {
  items: AIModel[]
  total: number
}

// 默认模型响应
export interface DefaultModelsResponse {
  text_generation?: AIModel
  image_generation?: AIModel
  image_analysis?: AIModel
}

// 连接测试结果
export interface AIModelTestResult {
  success: boolean
  message: string
  latency_ms?: number
}

// 提供商信息
export interface ProviderInfo {
  value: ModelProvider
  label: string
  supported_types: ModelType[]
}

// 模型类型信息
export interface ModelTypeInfo {
  value: ModelType
  label: string
}

/**
 * AI 模型管理 API
 */
export const aiModelApi = {
  /**
   * 获取模型列表
   */
  list: (params?: {
    model_type?: ModelType
    provider?: ModelProvider
    is_enabled?: boolean
  }): Promise<AIModelListResponse> => 
    api.get('/ai-models', { params }),

  /**
   * 获取各类型的默认模型
   */
  getDefaults: (): Promise<DefaultModelsResponse> => 
    api.get('/ai-models/defaults'),

  /**
   * 获取模型详情
   */
  get: (id: string): Promise<AIModel> => 
    api.get(`/ai-models/${id}`),

  /**
   * 创建模型配置
   */
  create: (data: AIModelCreateRequest): Promise<AIModel> => 
    api.post('/ai-models', data),

  /**
   * 更新模型配置
   */
  update: (id: string, data: AIModelUpdateRequest): Promise<AIModel> => 
    api.patch(`/ai-models/${id}`, data),

  /**
   * 删除模型配置
   */
  delete: (id: string): Promise<{ status: string; id: string }> => 
    api.delete(`/ai-models/${id}`),

  /**
   * 测试模型连接
   */
  test: (id: string): Promise<AIModelTestResult> => 
    api.post(`/ai-models/${id}/test`),

  /**
   * 设置为默认模型
   */
  setDefault: (id: string): Promise<AIModel> => 
    api.post(`/ai-models/${id}/set-default`),

  /**
   * 获取支持的提供商列表
   */
  getProviders: (): Promise<{ providers: ProviderInfo[] }> => 
    api.get('/ai-models/providers/list'),

  /**
   * 获取支持的模型类型列表
   */
  getModelTypes: (): Promise<{ types: ModelTypeInfo[] }> => 
    api.get('/ai-models/types/list'),
}

/**
 * 提供商显示名称映射
 */
export const providerLabels: Record<ModelProvider, string> = {
  openai: 'OpenAI',
  openai_compatible: 'OpenAI 兼容 (Azure/Ollama/vLLM)',
  anthropic: 'Anthropic Claude',
  stable_diffusion: 'Stable Diffusion WebUI',
  comfyui: 'ComfyUI',
  midjourney: 'Midjourney',
}

/**
 * 模型类型显示名称映射
 */
export const modelTypeLabels: Record<ModelType, string> = {
  text_generation: '文本生成',
  image_generation: '图像生成',
  image_analysis: '图像分析',
}

/**
 * 根据提供商获取需要的配置字段
 */
export const getFieldsByProvider = (provider: ModelProvider): string[] => {
  const common = ['name', 'model_name', 'timeout']
  
  switch (provider) {
    case 'openai':
      return [...common, 'api_key', 'base_url', 'organization_id']
    case 'openai_compatible':
      return [...common, 'api_key', 'base_url', 'api_version', 'deployment_name', 'extra_headers']
    case 'anthropic':
      return [...common, 'api_key', 'base_url', 'api_version']
    case 'stable_diffusion':
      return [...common, 'base_url', 'checkpoint_name', 'default_sampler', 
              'default_steps', 'default_cfg_scale', 'default_width', 'default_height']
    case 'comfyui':
      return [...common, 'base_url', 'workflow_template']
    case 'midjourney':
      return [...common, 'api_key', 'base_url']
    default:
      return common
  }
}

/**
 * 根据提供商获取支持的模型类型
 */
export const getSupportedTypesByProvider = (provider: ModelProvider): ModelType[] => {
  switch (provider) {
    case 'openai':
      return ['text_generation', 'image_generation', 'image_analysis']
    case 'openai_compatible':
    case 'anthropic':
      return ['text_generation', 'image_analysis']
    case 'stable_diffusion':
    case 'comfyui':
    case 'midjourney':
      return ['image_generation']
    default:
      return []
  }
}

File: backend/app/schemas/ai_model.py
"""AI模型配置相关Schema"""

from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, model_validator

from app.models.ai_model import ModelProvider, ModelType


class AIModelCreate(BaseModel):
    """创建模型配置请求"""
    # 基础字段
    name: str
    provider: ModelProvider
    model_type: ModelType
    model_name: str

    # 连接配置
    api_key: Optional[str] = None
    base_url: Optional[str] = None
    api_version: Optional[str] = None
    organization_id: Optional[str] = None
    deployment_name: Optional[str] = None

    # 文本生成配置
    default_max_tokens: Optional[int] = None
    default_temperature: Optional[float] = None

    # 图像生成配置
    default_width: Optional[int] = None
    default_height: Optional[int] = None
    default_steps: Optional[int] = None
    default_cfg_scale: Optional[float] = None
    default_sampler: Optional[str] = None
    checkpoint_name: Optional[str] = None
    workflow_template: Optional[str] = None

    # 通用配置
    timeout: int = 300
    extra_headers: Optional[Dict[str, str]] = None
    extra_config: Optional[Dict[str, Any]] = None

    is_enabled: bool = True
    is_default: bool = False

    @model_validator(mode='after')
    def validate_provider_fields(self):
        """根据供应商类型验证必需字段"""
        if self.provider in [ModelProvider.openai, ModelProvider.anthropic]:
            if not self.api_key:
                raise ValueError(f"{self.provider.value} 需要提供 api_key")
        if self.provider == ModelProvider.openai_compatible:
            if not self.api_key or not self.base_url:
                raise ValueError("openai_compatible 需要提供 api_key 和 base_url")
        if self.provider in [ModelProvider.stable_diffusion, ModelProvider.comfyui]:
            if not self.base_url:
                raise ValueError(f"{self.provider.value} 需要提供 base_url")
        if self.provider == ModelProvider.midjourney:
            if not self.api_key or not self.base_url:
                raise ValueError("midjourney 需要提供 api_key 和 base_url")
        return self


class AIModelUpdate(BaseModel):
    """更新模型配置请求 - 所有字段可选"""
    name: Optional[str] = None
    model_name: Optional[str] = None
    api_key: Optional[str] = None
    base_url: Optional[str] = None
    api_version: Optional[str] = None
    organization_id: Optional[str] = None
    deployment_name: Optional[str] = None

    default_max_tokens: Optional[int] = None
    default_temperature: Optional[float] = None

    default_width: Optional[int] = None
    default_height: Optional[int] = None
    default_steps: Optional[int] = None
    default_cfg_scale: Optional[float] = None
    default_sampler: Optional[str] = None
    checkpoint_name: Optional[str] = None
    workflow_template: Optional[str] = None

    timeout: Optional[int] = None
    extra_headers: Optional[Dict[str, str]] = None
    extra_config: Optional[Dict[str, Any]] = None

    is_enabled: Optional[bool] = None
    is_default: Optional[bool] = None


class AIModelResponse(BaseModel):
    """模型配置响应"""
    id: str
    name: str
    provider: ModelProvider
    model_type: ModelType
    model_name: str

    # 连接配置（不返回 api_key）
    base_url: Optional[str] = None
    api_version: Optional[str] = None
    organization_id: Optional[str] = None
    deployment_name: Optional[str] = None
    has_api_key: bool  # 只返回是否配置了密钥

    # 文本生成配置
    default_max_tokens: Optional[int] = None
    default_temperature: Optional[float] = None

    # 图像生成配置
    default_width: Optional[int] = None
    default_height: Optional[int] = None
    default_steps: Optional[int] = None
    default_cfg_scale: Optional[float] = None
    default_sampler: Optional[str] = None
    checkpoint_name: Optional[str] = None
    workflow_template: Optional[str] = None

    # 通用配置
    timeout: int
    extra_headers: Optional[Dict[str, str]] = None
    extra_config: Optional[Dict[str, Any]] = None

    is_enabled: bool
    is_default: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AIModelListResponse(BaseModel):
    """模型列表响应"""
    items: List[AIModelResponse]
    total: int


class AIModelTestResult(BaseModel):
    """模型连接测试结果"""
    success: bool
    message: str
    latency_ms: Optional[float] = None


class DefaultModelsResponse(BaseModel):
    """各类型默认模型响应"""
    text_generation: Optional[AIModelResponse] = None
    image_generation: Optional[AIModelResponse] = None
    image_analysis: Optional[AIModelResponse] = None
